lisa/tests/base.py

```diff
@@ -456,4 +457,42 @@ def _from_target(cls, target, res_dir, ftrace_coll=None):
 
         return cls(res_dir, plat_info, rtapp_profile)
 
+# TODO: move this in the RTATestBundle class?
+def check_noisy_tasks(threshold_pct=5):
+    def decorator(func):
+        @functools.wraps(func)
+        def wrapper(self, *args, **kwargs):
+            df = self.trace.analysis.tasks.df_tasks_runtime()
+
+            test_tasks = list(map(self.trace.get_task_pid, self.rtapp_profile.keys()))
+
+            # TODO: parametize this
+            # Don't count the idle task as noise
+            df_noise = df[~df.index.isin(test_tasks + [0])]
+
+            if not df_noise.empty:
+                pid = df_noise.index[0]
+                comm = df_noise.comm.values[0]
+                duration = df_noise.runtime.values[0]
+                duration_pct = 100 * duration / self.trace.time_range
+                ok = duration_pct < threshold_pct
+
+                def checker():
+                    res = func(self, *args, **kwargs)
+
+                    if not ok:
+                        res.result = Result.UNDECIDED
+
+                    metric = {"pid" : pid,
+                              "comm": comm,
+                              "duration": TestMetric(duration, "s"),
+                              "duration_pct" : TestMetric(duration_pct, "%")}
+                    res.add_metric("noisiest task", metric)
```

Comment:
s/noisiest/noisy/?

You're not filtering for the noisiest and just creating a list of noisy tasks, right?

Reply:
No we are actually just picking the noisiest one. There's no need to list all of them (there can be plenty), just the worst offender.

Reply:
OK so the list is sorted and index[0] contains the task with the highest runtime value?

It depends on the sequence of events but it can happen that the noisiest ran during 'idle' time of the rt-app task but the 2nd busiest is the culprit, no? I don't expect the list of noisy task to be high, have you seen a lot? Was that on Debian or Buildroot? In my testing previously (Debian) I saw 3 or 4 max IIRC.

Reply:
> OK so the list is sorted and index[0] contains the task with the highest runtime value?
> 

Yeah the dataframe is sorted on runtime values, so any value at index 0 will be the one with the highest runtime.

> It depends on the sequence of events but it can happen that the noisiest ran during 'idle' time of the rt-app task but the 2nd busiest is the culprit, no? I don't expect the list of noisy task to be high, have you seen a lot? Was that on Debian or Buildroot? In my testing previously (Debian) I saw 3 or 4 max IIRC.

Right, tasks running during the "active" part of our rt-app workload would have a higher impact, but I think focusing on the longest-running (aka noisiest) undesired task is a safe proxy. It gives us a "worst case scenario" value that we can easily compare against a threshold (which ATM is a % of the total execution time of our rt-app workload).

Also, even if the tasks run while the rt-app tasks are sleeping, they can affect the test in some way or another: frequency increase we wouldn't see otherwise, unexpected load/util increases (w/ blocked load that could impact the rt-app workload even if it's not being coscheduled with the noisy task), etc.

Reply:
>Also, even if the tasks run while the rt-app tasks are sleeping, they can affect the test in some way or another: frequency increase we wouldn't see otherwise, unexpected load/util increases (w/ blocked load that could impact the rt-app workload even if it's not being coscheduled with the noisy task), etc.

This further confirms that the timing of when the task run is as important as the duration. But I am happy with this as is if that's what you feel is the best for now. We can improve on this as we go along of we need to :-)

Reply:
I'm curious as to what you have in mind - would you rather do something like check the noisiness in each time window where we have an active rt-app task (thus ignoring time windows where we are not running any rt-app task, but are still between first & last rt-app events)?

I do believe taking idle times into account is just as important (as I've explained above). I also have an extra example:

I still haven't ported to our test framework (only as notebook), but I have a test that makes sure we correctly decay nohz blocked load (here's some blurb about it: https://lore.kernel.org/lkml/b1efd041-2faa-76e0-7d9b-3d2917e03012@arm.com/). We *really* don't want anything to run during that decay phase, and during this phase we don't have any rt-app task running (since we want to observe this decay). Arguably that's a bit of a special case, but it does help back the way I've done this check.

Reply:
No what I was saying is that the noisiest doesn't necessarily correlate to the problematic interfering task. Another noisy task that happens to run at the wrong time but for a much shorter period could be the one that break assumptions. e.g: a kworker waking up at the same time with another task for 20us but causing the rt-app task to migrate. In the same test there could be another noisy task that ran for 0.2 ms which you'd detect as the noisiest but it's completely harmless.

So what I had in mind is keeping a list of how many noisy tasks were there in total and possibly there names/pids to make it easier to look for them in kernelshark. Noisy task is more than just about duration and while I don't think it's worth doing complex checks to define noisy but keeping a full list is helpful IMO as there can be several modes of interference.

Reply:
Right, you've got a point. FWIW, if you execute `trace.analysis.tasks.df_tasks_runtime()` you'll get that list (comm & pid) sorted by total runtime (the noise test just filters that dataframe). It doesn't help for the test but it can help to figure out what to look for in kernelshark

Reply:
Theses metrics are primarily meant for quick inspection of the `ResultBundle` (`-oshow-details` in bisector will show them for example). If we can think of something that is better for human consumption and that would not provide huge output (if that cannot be avoided), it may become a better candidate as a metric. As Valentin pointed, the data are there in any case if somebody needs to sort out the wheat from chaff on a particularly bad set of `TestBundle`.

Reply:
>As Valentin pointed, the data are there in any case if somebody needs to sort out the wheat from chaff on a particularly bad set of TestBundle.

We already do this :-)

Anyway my point was that we need to be aware that noisiest doesn't necessarily correlate to problematic. As long as we agree on that and ready to react to this as experience teaches us more I have no issue with the current implementation. It's all good stuff :-)

FWIW when I had this noise detection code in the past I did encounter cases where aggregated sum of noisy tasks was causing the problem. But I think with buildroot where things are quiet this is hard to happen and generally I expect a very small number of noisy tasks. If it's large then this is a separate issue that needs investigating.

Reply:
> FWIW when I had this noise detection code in the past I did encounter cases where aggregated sum of noisy tasks was causing the problem

Right, if we start seeing that issue as well on some setups, it should be straightforward to also check the sum. At this point we can add optional parameters to the decorator to select the flavor of filtering that is more appropriate for a given test. The current decorator implementation is not trivial, but it brings a high level of flexibility in specifying defaults and letting the caller of the test override these defaults :-)